raft: Fix infinite election loop

aaronlehmann opened the pull request:
It looks like the election loops on startup are caused by something
proposing a value before all entries in the log have become committed.
This would cause ApplyStoreActions to try and get a lock on the raft
store, while processInternalRaftRequest is holding the lock (since it's
called by the memory store with the lock held). This situation is a
deadlock. It blocks the loop in Node.Run and prevents further
participation in the cluster.

To try and solve the issue, don't signal that the local node has become
the leader until it has committed all the uncommitted entries from
before. Also, block proposals until it is caught up.

WIP because we still need to test and confirm that this fixes the problem.

LK4D4 commented:
@aaronlehmann looks like real CI failure.
Do you have reproduction case? I can try with real machines.

aaronlehmann commented:
Tests passed locally for me. Looks like the failure in CI is because the test needs an adjustment now that the conditions for proposing a value are stricter. I'll work on fixing the test tomorrow.

I think the random failures we saw in certain tests were related to this problem.

To reproduce, set up a cluster of three nodes, then kill dockerd on all three, and restart the old leader and one of the followers at the same time. Repeat the restart of the two nodes until it gets stuck.

aaronlehmann commented:
I think I fixed the tests. Let's see what happens in CI.

I also added a commit that prevents a follower from sending a proposal, or the leader from accepting one. Only the leader should make proposals, so we avoid ever having conflicting proposals. There is already code to prevent this, but there corner cases where the leader could lose its leadership status right while it's proposing a value.

aaronlehmann commented:
Vendored this in Docker and it passes the integration tests.

LK4D4 commented:
ping @abronan

abronan commented:
Removed the WIP and moved to code review.

LGTM

LK4D4 commented:
LGTM
